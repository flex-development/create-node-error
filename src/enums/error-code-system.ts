/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
