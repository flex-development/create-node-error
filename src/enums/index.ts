/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
