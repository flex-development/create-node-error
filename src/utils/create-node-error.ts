/**
 * @file Utilities - createNodeError
 * @module errnode/utils/createNodeError
 * @see https://github.com/nodejs/node/blob/v19.3.0/lib/internal/errors.js#L367-L404
 */

import type { ErrorCode } from '#src/enums'
import type { NodeErrorConstructor } from '#src/interfaces'
import type { Args } from '#src/interfaces/node-error-constructor'
import kIsNodeError from '#src/internal/k-is-node-error'
import type { ErrnodeConstructor, Message, MessageFn } from '#src/types'
import { cast, define, isString, regexp } from '@flex-development/tutils'
import { format } from 'node-inspect-extracted'

/**
 * Creates a Node.js error constructor.
 *
 * Constructor arguments are passed to [`util.format`][1] if the error `message`
 * is a string, or `message` itself if it is a function. Message functions will
 * also be called with the new Node.js error instance as `this`.
 *
 * [1]: https://nodejs.org/api/util.html#utilformatformat-args
 *
 * @see {@linkcode NodeErrorConstructor}
 * @see https://nodejs.org/api/util.html#utilformatformat-args
 *
 * @template B - Error base class type
 * @template M - Error message function or `util.format` arguments
 * @template T - Error prototype
 *
 * @param {ErrorCode} code - Node.js error code
 * @param {B} Base - Error base class
 * @param {Message<M>} message - Error message string or function
 * @return {NodeErrorConstructor<T, M>} `NodeError` constructor
 */
const createNodeError = <
  B extends ErrnodeConstructor,
  M extends MessageFn | string | readonly unknown[],
  T extends B['prototype']
>(
  code: ErrorCode,
  Base: B,
  message: Message<M>
): NodeErrorConstructor<T, M> => {
  // @ts-expect-error ts(2322)
  return class NodeError extends Base {
    /**
     * Node.js error code.
     *
     * @public
     * @instance
     * @member {ErrorCode}
     */
    public code!: ErrorCode

    /**
     * Creates a new Node.js error.
     *
     * @param {any[]} args - Error message function arguments
     */
    constructor(...args: any[]) {
      super()

      /**
       * Instance properties are defined in order of precedence.
       *
       * Precedence is determined by latter property definitions using the value
       * of former property definitions to generate a value.
       *
       * Instance property dependencies:
       *
       * - `message`: `code`
       * - `toString`: `code`, `message`, `name`
       * - `stack`: `toString`
       */

      // add error symbol
      define(this, kIsNodeError, {
        configurable: true,
        enumerable: false,
        value: true,
        writable: false
      })

      // add error code
      define(this, 'code', {
        configurable: true,
        enumerable: true,
        value: code,
        writable: true
      })

      // add error message
      define(this, 'message', {
        configurable: true,
        enumerable: false,
        value: NodeError.#message(code, message, cast(args), this),
        writable: true
      })

      // redefine toString method
      define(this, 'toString', {
        configurable: true,
        enumerable: false,
        /**
         * Returns a string representation of the error.
         *
         * @this {import('#src/types').NodeError}
         *
         * @return {string} String representation of error
         */
        value(this: NodeError): string {
          return `${this.name} [${this.code}]: ${this.message}`
        },
        writable: true
      })

      // add stack trace
      define(this, 'stack', {
        configurable: true,
        enumerable: true,
        value: NodeError.#prepareStackTrace(this),
        writable: true
      })
    }

    /**
     * Creates an error message string.
     *
     * @see https://nodejs.org/api/util.html#utilformatformat-args
     *
     * @private
     * @static
     *
     * @param {ErrorCode} code - Node.js error code
     * @param {Message<M>} message - Error message function or format string
     * @param {Args<M>} args - Custom message function arguments if `message` is
     * a function, or `util.format` arguments if `message` is a string
     * @param {Error} self - Error object used as `this` argument
     * @return {string} Formatted error message
     * @throws {Error} If `args` length is invalid
     */
    static #message(
      code: ErrorCode,
      message: Message<M>,
      args: Args<M>,
      self: Error
    ): string {
      /**
       * Invalid {@linkcode args} length error message.
       *
       * @var {string} error
       */
      let error: string = `${code};` + ' '

      // check args.length against expected message function arguments length
      // default parameters do not count when message is a function
      if (isString(message)) {
        /**
         * {@linkcode format} specifiers regex.
         *
         * @const {RegExp} regex
         */
        const regex: RegExp = /%[Odfijos]/g

        /**
         * Expected message length.
         *
         * @var {number} length
         */
        let length: number = 0

        // get expected message length
        while (regex.exec(message) !== null) length++

        // throw if incorrect number of arguments are passed to util.format
        if (length !== args.length) {
          error += `The arguments length (${args.length}) provided to \`util.format\` does not match the required length (${length}).`
          throw new Error(error, { cause: args })
        }
      } else if (!(message.length <= args.length)) {
        error += `The arguments length (${args.length}) provided to \`message\` does not match the required length (${message.length}).`
        throw new Error(error, { cause: args })
      }

      return Reflect.apply(
        isString(message) ? format : message,
        isString(message) ? null : self,
        isString(message) ? [message, ...args] : args
      )
    }

    /**
     * Adds a formatted stack trace to the given `error`.
     *
     * Previously written stack traces are not preserved.
     *
     * @see https://v8.dev/docs/stack-trace-api#customizing-stack-traces
     *
     * @private
     * @static
     *
     * @param {Error} error - Error object to prepare stack trace for
     * @return {string} Formatted stack trace
     */
    static #prepareStackTrace(error: Error): string {
      // create stack property
      Base.captureStackTrace(error, NodeError.#prepareStackTrace)

      // format stack trace
      return error.stack!.replace(
        new RegExp(`^${error.name}: ${regexp(error.message)}`),
        error.toString()
      )
    }
  }
}

export default createNodeError
