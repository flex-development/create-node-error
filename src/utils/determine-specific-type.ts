/**
 * @file Utilities - determineSpecificType
 * @module errnode/utils/determineSpecificType
 * @see https://github.com/nodejs/node/blob/v19.3.0/lib/internal/errors.js#L878-L896
 */

import { cast } from '@flex-development/tutils'
import { inspect } from 'node-inspect-extracted'

/**
 * Determines the specific type of a value for type-mismatch errors.
 *
 * @param {unknown} value - Value to evaluate
 * @return {string} Specific type of `value`
 */
function determineSpecificType(value: unknown): string {
  /**
   * Specific type of `value`.
   *
   * @var {string} type
   */
  let type: string = ''

  switch (true) {
    case typeof value === 'function':
      type = `function ${cast<FunctionConstructor>(value).name}`
      break
    case typeof value === 'object':
      type = value?.constructor?.name
        ? `an instance of ${value.constructor.name}`
        : inspect(value, { depth: -1 })
      break
    case typeof value === 'undefined':
      type = typeof value
      break
    default:
      /**
       * String representation of {@linkcode value}.
       *
       * @var {string} inspected
       */
      let inspected: string = inspect(value, { colors: false })

      // trim string representation of value
      if (inspected.length > 28) inspected = inspected.slice(0, 25) + '...'

      type = `type ${typeof value} (${inspected})`
      break
  }

  return type
}

export default determineSpecificType
