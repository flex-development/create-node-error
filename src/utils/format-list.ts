/**
 * @file Utilities - formatList
 * @module errnode/utils/formatList
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1043-L1060
 */

import { at, join } from '@flex-development/tutils'

/**
 * Create a list string in the form `'A and B'` or `'A, B, ..., and Z`.
 *
 * > {@linkcode Intl.ListFormat} cannot be used because it's not available in
 * > `--without-intl` builds.
 *
 * @example
 *  formatList([])
 *  // ''
 * @example
 *  formatList(['json'])
 *  // 'json'
 * @example
 *  formatList(['data', 'file'])
 *  // 'data and file'
 * @example
 *  formatList(['boolean', 'number', 'string'], 'or')
 *  // 'boolean, number, or string'
 *
 * @param {ReadonlyArray<string> | string} list
 *  List to format
 * @param {'and' | 'or' | null | undefined} [type='and']
 *  List transition type to insert before last element
 * @return {string}
 *  Formatted list
 */
function formatList(
  list: string | readonly string[],
  type?: 'and' | 'or' | null | undefined
): string {
  if (typeof list === 'string') return list

  if (!type) type = 'and'

  switch (list.length) {
    case 0:
      return ''
    case 1:
      return `${list[0]}`
    case 2:
      return `${list[0]} ${type} ${list[1]}`
    case 3:
      return `${list[0]}, ${list[1]}, ${type} ${list[2]}`
    default:
      return `${join(list.slice(0, -1), ', ')}, ${type} ${at(list, -1)}`
  }
}

export default formatList
