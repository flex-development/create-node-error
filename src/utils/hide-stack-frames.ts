/**
 * @file Utilities - hideStackFrames
 * @module errnode/utils/hideStackFrames
 */

import type { Fn } from '@flex-development/tutils'

/**
 * Remove unnecessary frames from Node.js core errors.
 *
 * @template {Fn} T
 *  Function to wrap
 *
 * @param {T} fn
 *  Function to wrap
 * @return {T}
 *  `fn`
 */
function hideStackFrames<T extends Fn>(fn: T): T {
  // rename functions that will be hidden.
  // this cuts off the stack trace at the outermost one.
  Object.defineProperty(fn, 'name', { value: '__node_internal_' + fn.name })
  return fn
}

export default hideStackFrames
