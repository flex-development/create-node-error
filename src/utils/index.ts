/**
 * @file Entry Point - Utilities
 * @module errnode/utils
 */

export { default as createNodeError } from './create-node-error'
export { default as determineSpecificType } from './determine-specific-type'
