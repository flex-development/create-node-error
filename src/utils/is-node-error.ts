/**
 * @file Utilities - isNodeError
 * @module errnode/utils/isNodeError
 */

import kIsNodeError from '#k-is-node-error'
import type { NodeError } from '#src/interfaces'

/**
 * Check if `value` is a {@linkcode NodeError}.
 *
 * @param {unknown} value
 *  Value to check
 * @return {value is NodeError}
 *  `true` if `value` is `NodeError`, `false` otherwise
 */
function isNodeError(value: unknown): value is NodeError {
  return (
    typeof value === 'object' &&
    value !== null &&
    value instanceof Error &&
    kIsNodeError in value &&
    value[kIsNodeError] === true
  )
}

export default isNodeError
