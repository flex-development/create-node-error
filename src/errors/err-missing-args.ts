/**
 * @file Errors - ERR_MISSING_ARGS
 * @module errnode/errors/ERR_MISSING_ARGS
 * @see https://github.com/nodejs/node/blob/v22.8.0/lib/internal/errors.js#L1565-L1579
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import { formatList } from '#src/utils'
import { ok } from 'devlop'

/**
 * `ERR_MISSING_ARGS` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_missing_args
 *
 * @extends {NodeError<codes.ERR_MISSING_ARGS>}
 * @extends {TypeError}
 */
interface ErrMissingArgs extends NodeError<codes.ERR_MISSING_ARGS>, TypeError {}

/**
 * `ERR_MISSING_ARGS` message arguments.
 */
type Args = [name: string | string[], ...names: (string | string[])[]]

/**
 * `ERR_MISSING_ARGS` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrMissingArgs}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrMissingArgs,Args>}
 */
interface ErrMissingArgsConstructor
  extends NodeErrorConstructor<ErrMissingArgs, Args> {
  /**
   * Create a new `ERR_MISSING_ARGS` error.
   *
   * @see {@linkcode ErrMissingArgs}
   *
   * @param {string} name
   *  Name of missing argument
   * @param {(string | string[])[]} names
   *  List of missing arguments
   * @return {ErrMissingArgs}
   */
  new (name: string, ...names: (string | string[])[]): ErrMissingArgs
}

/**
 * `ERR_MISSING_ARGS` model.
 *
 * Thrown when a required argument of a Node.js API was not passed.
 *
 * @see {@linkcode ErrMissingArgsConstructor}
 *
 * @type {ErrMissingArgsConstructor}
 *
 * @class
 */
const ERR_MISSING_ARGS: ErrMissingArgsConstructor = E(
  codes.ERR_MISSING_ARGS,
  TypeError,
  /**
   * @param {(string | string[])[]} names
   *  List of missing arguments
   * @return {string}
   *  Error message
   */
  function message(...names: (string | string[])[]): string {
    ok(names.length, 'At least one arg needs to be specified')

    /**
     * Wrap a string in quotes.
     *
     * @param {string} a
     *  String to wrap
     * @return {string}
     *  Wrapped string
     */
    const wrap = (a: string): string => `"${a}"`

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = 'The '

    message += formatList(names.map(a => {
      /* c8 ignore next 2 */ return Array.isArray(a)
        ? a.map(wrap).join(' or ')
        : wrap(a)
    }))

    return `${message} argument${names.length > 1 ? 's' : ''} must be specified`
  }
)

export {
  ERR_MISSING_ARGS as default,
  type ErrMissingArgs,
  type Args as ErrMissingArgsArgs,
  type ErrMissingArgsConstructor
}
