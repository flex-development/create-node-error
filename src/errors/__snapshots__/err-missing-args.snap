// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_MISSING_ARGS > should build message from parameters (0) 1`] = `"The "domain" argument must be specified"`;

exports[`unit:errors/ERR_MISSING_ARGS > should build message from parameters (1) 1`] = `"The "name" and "value" arguments must be specified"`;
